src/hotspot/share/compiler/compilerDefinitions.cpp

```diff
@@ -38,47 +41,72 @@ const char* compilertype2name_tab[compiler_number_of_types] = {
   "jvmci"
 };
 
-#ifdef TIERED
-bool CompilationModeFlag::_quick_only = false;
-bool CompilationModeFlag::_high_only = false;
-bool CompilationModeFlag::_high_only_quick_internal = false;
+CompilationModeFlag::Mode CompilationModeFlag::_mode = CompilationModeFlag::Mode::NORMAL;
 
+static void print_mode_unavailable(const char* mode_name, const char* reason) {
+  warning("%s compilation mode unavailable because %s.", mode_name, reason);
+}
 
 bool CompilationModeFlag::initialize() {
+  _mode = Mode::NORMAL;
   if (CompilationMode != NULL) {
-    if (strcmp(CompilationMode, "default") == 0) {
-      // Do nothing, just support the "default" keyword.
+    if (strcmp(CompilationMode, "default") == 0 || strcmp(CompilationMode, "normal") == 0) {
+      assert(_mode == Mode::NORMAL, "Precondition");
     } else if (strcmp(CompilationMode, "quick-only") == 0) {
-      _quick_only = true;
+      if (!CompilerConfig::has_c1()) {
+        print_mode_unavailable("quick-only", "there is no c1 present");
+      } else {
+        _mode = Mode::QUICK_ONLY;
+      }
     } else if (strcmp(CompilationMode, "high-only") == 0) {
-      _high_only = true;
+      if (!CompilerConfig::has_c2() && !CompilerConfig::is_jvmci_compiler()) {
```

Comment:
Is using `!is_c2_or_jvmci_compiler_available()` better here? All flags should be processed at this point. And we could have some `TieredStopAtLevel` flags.
It looks like you have cross dependency between CompilerConfig and CompilationModeFlag which make using `is_c2_or_jvmci_compiler_available()` impossible here.

Reply:
Yes, there is an unfortunate cross dependency. It would probably work fine since ```_mode``` is initialized to ```NORMAL``` before parsing the flag. But I wanted to not use any function that would depend on the value of the ```_mode``` during parsing. It could create trouble in the future. I'll put more comments about that.

Reply:
okay.